src/embed/http.rs

```diff
@@ -1,96 +1,234 @@
 use std::convert::TryFrom;
-use std::error::Error;
-use std::fmt;
-use std::str;
 use wasm_bindgen::JsCast;
 use wasm_bindgen::JsValue;
 use wasm_bindgen_futures::JsFuture;
 use web_sys::{RequestInit, RequestMode};
 
+// s makes map_err calls nicer by mapping a error to its debug-printed string.
+fn s<D: std::fmt::Debug>(err: D) -> String {
+    format!("{:?}", err)
+}
+
+// rust_fetch fetches using a rust (as opposed to browser Fetch) http client.
+// It consumes the input request (it needs to own the request body). The response
+// returned will have status and body filled (doesn't do headers at this point).
+// Non-200 status code does not constitute an Err Result.
+//
+// TODO TLS
+// TODO timeout/abort
+// TODO log req/resp
+// TODO understand what if any tokio runtime assumptions are implied here
+pub async fn rust_fetch(
+    hyper_client: &hyper::Client<hyper::client::HttpConnector>,
+    http_req: http::Request<String>,
+) -> Result<http::Response<String>, FetchError> {
+    let (parts, req_body) = http_req.into_parts();
+    let mut builder = hyper::Request::builder()
+        .method(parts.method.as_str())
+        .uri(&parts.uri.to_string());
+    for (k, v) in parts.headers.iter() {
+        builder = builder.header(
+            k,
+            v.to_str()
+                .map_err(|e| FetchError::InvalidRequestHeader(s(e)))?,
```

Comment:
[help] I don't understand why `s()` is needed. When I implemented this pattern myself, I found that a nice string was automatically generated, recursively, when I tried to print the top level error. So the functionality in `s` was not necessary.

Doing it this way you are discards other information in `e` -- its type, which is a big part of why I liked this pattern. I'm not in favor of it if we have to do this.

Perhaps this is where the `std::err::Error` trait comes in?

Reply:
this was an explicit choice i call out in the comment bc there were multiple lower level types. i agree that using std::err:error is better, not sure why i didn't do that. anyway created a follow up: https://github.com/rocicorp/repc/issues/83